pkg/plugin/framework/item_snapshotter_server.go

```diff
@@ -0,0 +1,299 @@
+	if err != nil {
+		return nil, newGRPCError(err)
+	}
+	var backup api.Backup
+
+	if err := json.Unmarshal(req.Backup, &backup); err != nil {
+		return nil, newGRPCError(errors.WithStack(err))
+	}
+	sipi := &isv1.ProgressInput{
+		ItemID:     protoToResourceIdentifier(req.ItemID),
+		SnapshotID: req.SnapshotID,
+		Backup:     &backup,
+	}
+
+	sipo, err := impl.Progress(sipi)
+	if err != nil {
+		return nil, newGRPCError(err)
+	}
+	res = &proto.ProgressResponse{
+		Completed:       sipo.Completed,
+		ItemsCompleted:  sipo.ItemsCompleted,
+		ItemsToComplete: sipo.ItemsToComplete,
+		Started:         sipo.Started.Unix(),
+		StartedNano:     sipo.Started.UnixNano(),
+		Updated:         sipo.Updated.Unix(),
+		UpdatedNano:     sipo.Updated.UnixNano(),
+	}
+	return res, nil
+}
+
+func (recv *ItemSnapshotterGRPCServer) DeleteSnapshot(ctx context.Context, req *proto.DeleteItemSnapshotRequest) (empty *proto.Empty, err error) {
+	defer func() {
+		if recoveredErr := handlePanic(recover()); recoveredErr != nil {
+			err = recoveredErr
+		}
+	}()
+	impl, err := recv.getImpl(req.Plugin)
+	if err != nil {
+		return nil, newGRPCError(err)
+	}
+
+	var itemFromBackup unstructured.Unstructured
+	if err := json.Unmarshal(req.ItemFromBackup, &itemFromBackup); err != nil {
+		return nil, newGRPCError(errors.WithStack(err))
+	}
+
+	disi := isv1.DeleteSnapshotInput{
+		SnapshotID:       req.SnapshotID,
+		ItemFromBackup:   &itemFromBackup,
+		SnapshotMetadata: req.Metadata,
+		Params:           req.Params,
+	}
+
+	err = impl.DeleteSnapshot(ctx, &disi)
+	if err != nil {
+		return nil, newGRPCError(err)
+	}
+	return
+}
+
+func (recv *ItemSnapshotterGRPCServer) CreateItemFromSnapshot(ctx context.Context, req *proto.CreateItemFromSnapshotRequest) (res *proto.CreateItemFromSnapshotResponse, err error) {
+	defer func() {
+		if recoveredErr := handlePanic(recover()); recoveredErr != nil {
+			err = recoveredErr
+		}
+	}()
+	impl, err := recv.getImpl(req.Plugin)
+	if err != nil {
+		return nil, newGRPCError(err)
+	}
+
+	var snapshottedItem unstructured.Unstructured
+	if err := json.Unmarshal(req.Item, &snapshottedItem); err != nil {
+		return nil, newGRPCError(errors.WithStack(err))
+	}
+
+	var itemFromBackup unstructured.Unstructured
+	if err := json.Unmarshal(req.Item, &itemFromBackup); err != nil {
+		return nil, newGRPCError(errors.WithStack(err))
+	}
+
+	var restore api.Restore
+
+	if err := json.Unmarshal(req.Restore, &restore); err != nil {
+		return nil, newGRPCError(errors.WithStack(err))
+	}
+
+	cii := isv1.CreateItemInput{
+		SnapshottedItem:  &snapshottedItem,
+		SnapshotID:       req.SnapshotID,
+		ItemFromBackup:   &itemFromBackup,
+		SnapshotMetadata: req.SnapshotMetadata,
+		Params:           req.Params,
+		Restore:          &restore,
+	}
+
+	cio, err := impl.CreateItemFromSnapshot(ctx, &cii)
+	if err != nil {
+		return nil, newGRPCError(err)
+	}
+
+	res = &proto.CreateItemFromSnapshotResponse{
+		Item:            nil,
+		AdditionalItems: nil,
```

Comment:
Should `Item` and `AdditionalItem` be set to the fields from `cio`?

Reply:
Probably, yes